homeassistant/components/somfy/__init__.py

```diff
@@ -0,0 +1,162 @@
+"""
+Support for Somfy hubs.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/somfy/
+"""
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+from homeassistant import config_entries
+from homeassistant.components.somfy import config_flow
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_TOKEN
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import HomeAssistantType
+from homeassistant.util import Throttle
+
+API = 'api'
+
+DEVICES = 'devices'
+
+REQUIREMENTS = ['pymfy==0.5.1']
+
+_LOGGER = logging.getLogger(__name__)
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=10)
+
+DOMAIN = 'somfy'
+
+CONF_CLIENT_ID = 'client_id'
+CONF_CLIENT_SECRET = 'client_secret'
+
+SOMFY_AUTH_CALLBACK_PATH = '/auth/somfy/callback'
+SOMFY_AUTH_START = '/auth/somfy'
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_CLIENT_ID): cv.string,
+        vol.Required(CONF_CLIENT_SECRET): cv.string
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+SOMFY_COMPONENTS = ['cover']
+
+
+async def async_setup(hass, config):
+    """Set up the Somfy component."""
+    if DOMAIN not in config:
+        return True
+
+    hass.data[DOMAIN] = {}
+
+    config_flow.register_flow_implementation(
+        hass, DOMAIN, config[DOMAIN][CONF_CLIENT_ID],
+        config[DOMAIN][CONF_CLIENT_SECRET])
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={'source': config_entries.SOURCE_IMPORT},
+        ))
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Set up Somfy from a config entry."""
+    def token_saver(token):
+        _LOGGER.debug('Saving updated token')
+        entry.data[CONF_TOKEN] = token
+        hass.config_entries.async_update_entry(entry, data={**entry.data})
+
+    # Force token update.
+    from pymfy.api.somfy_api import SomfyApi
+    hass.data[DOMAIN][API] = SomfyApi(
```

Comment:
Does the somfy api support asyncio?

Reply:
It's not allowed to do I/O in the event loop. Coroutines will be executed in the event loop. That's why I'm asking about "async safe" below.

Reply:
The whole library pymfy (wrapper around Somfy API) does not perform any I/O.

Reply:
Oauth requests is considered I/O. Were sending info over the network, so it's not pure memory operations.

Reply:
I've compared my library to https://github.com/fredrike/pypoint used by the Point component. We use the same library, and apparently pypoint does not any trouble with async.

Reply:
See https://github.com/home-assistant/home-assistant/issues/19312.

Reply:
Oh!

Reply:
If you're working on the library, I would recommend converting it to use asyncio now, rather than later. It will make things a lot easier. Probably try to use this client library:
https://github.com/klen/aioauth-client

Reply:
Thx for sharing this library. I can have a look to update my library. It seems to be a lot of work. Do you think, it's a good idea to have both behavior in the same module, sync and async? This way, end user can use what it want.

Reply:
Probably separate asyncio from requests by module but support for both could be kept in the same library. Supporting both will mean a lot more work and complexity though, if you want to be able to reuse some code for the different modes.

I would never recommend supporting both asyncio and non asyncio for a new library but since there already is support for requests you could decide to keep that.

Reply:
I've explored https://github.com/klen/aioauth-client, but it's not so yet complete (missing refresh url, and token updater). So update my library will take eternity. As said below, use hass.async_add_executor_job will be enough? If yes, later I can take time to create (at least try) a replacement of requests-oauthlib.

Reply:
Yes, we can use `hass.async_add_executor_job`.

Reply:
Fixed.